In [1]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

In [2]:
tf.random.set_seed(42)

In [3]:
epochs = 200
batch_size = 128
verbose = 1
nb_classes = 10
n_hidden = 128
validation_split = 0.2
dropout = 0.3

In [4]:
mnist = keras.datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [5]:
print(X_train.shape, X_test.shape)

(60000, 28, 28) (10000, 28, 28)


In [6]:
reshaped = 784

X_train = X_train.reshape(60000, reshaped)
X_test = X_test.reshape(10000, reshaped)

In [7]:
X_train = X_train.astype(float)
X_test = X_test.astype(float)

In [8]:
X_train = X_train / 255
X_test = X_test / 255

In [9]:
y_train = tf.keras.utils.to_categorical(y_train, nb_classes)
y_test = tf.keras.utils.to_categorical(y_test, nb_classes)

In [10]:
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(n_hidden,
                             input_shape=(reshaped,),
                             activation='relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(n_hidden,
                             activation='relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(nb_classes,
                             activation='softmax'))

2022-03-16 16:31:51.140584: I tensorflow/core/platform/cpu_feature_guard.cc:145] This TensorFlow binary is optimized with Intel(R) MKL-DNN to use the following CPU instructions in performance critical operations:  SSE4.1 SSE4.2
To enable them in non-MKL-DNN operations, rebuild TensorFlow with the appropriate compiler flags.
2022-03-16 16:31:51.141650: I tensorflow/core/common_runtime/process_util.cc:115] Creating new thread pool with default inter op setting: 8. Tune using inter_op_parallelism_threads for best performance.


In [11]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               16512     
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


In [12]:
model.compile(optimizer='SGD', loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
model.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=verbose,
          validation_split=validation_split)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 4s 76us/sample - loss: 1.7641 - accuracy: 0.4315 - val_loss: 0.9905 - val_accuracy: 0.7911
Epoch 2/200
48000/48000 [==============================] - 3s 62us/sample - loss: 0.9669 - accuracy: 0.7036 - val_loss: 0.5541 - val_accuracy: 0.8658
Epoch 3/200
48000/48000 [==============================] - 3s 63us/sample - loss: 0.7158 - accuracy: 0.7794 - val_loss: 0.4309 - val_accuracy: 0.8887
Epoch 4/200
48000/48000 [==============================] - 3s 61us/sample - loss: 0.6038 - accuracy: 0.8163 - val_loss: 0.3738 - val_accuracy: 0.8976
Epoch 5/200
48000/48000 [==============================] - 3s 61us/sample - loss: 0.5366 - accuracy: 0.8388 - val_loss: 0.3401 - val_accuracy: 0.9029
Epoch 6/200
48000/48000 [==============================] - 3s 62us/sample - loss: 0.4917 - accuracy: 0.8523 - val_loss: 0.3165 - val_accuracy: 0.9096
Epoch 7/200
48000/48000 [=========================

Epoch 55/200
48000/48000 [==============================] - 3s 59us/sample - loss: 0.1701 - accuracy: 0.9493 - val_loss: 0.1179 - val_accuracy: 0.9641
Epoch 56/200
48000/48000 [==============================] - 3s 59us/sample - loss: 0.1695 - accuracy: 0.9484 - val_loss: 0.1174 - val_accuracy: 0.9653
Epoch 57/200
48000/48000 [==============================] - 3s 59us/sample - loss: 0.1659 - accuracy: 0.9507 - val_loss: 0.1163 - val_accuracy: 0.9649
Epoch 58/200
48000/48000 [==============================] - 3s 58us/sample - loss: 0.1662 - accuracy: 0.9507 - val_loss: 0.1159 - val_accuracy: 0.9660
Epoch 59/200
48000/48000 [==============================] - 3s 59us/sample - loss: 0.1645 - accuracy: 0.9517 - val_loss: 0.1145 - val_accuracy: 0.9658
Epoch 60/200
48000/48000 [==============================] - 3s 60us/sample - loss: 0.1622 - accuracy: 0.9520 - val_loss: 0.1141 - val_accuracy: 0.9662
Epoch 61/200
48000/48000 [==============================] - 3s 60us/sample - loss: 0.1605 - ac

48000/48000 [==============================] - 3s 68us/sample - loss: 0.1120 - accuracy: 0.9660 - val_loss: 0.0904 - val_accuracy: 0.9743
Epoch 110/200
48000/48000 [==============================] - 3s 64us/sample - loss: 0.1095 - accuracy: 0.9676 - val_loss: 0.0901 - val_accuracy: 0.9742
Epoch 111/200
48000/48000 [==============================] - 3s 70us/sample - loss: 0.1096 - accuracy: 0.9665 - val_loss: 0.0901 - val_accuracy: 0.9740
Epoch 112/200
48000/48000 [==============================] - 3s 65us/sample - loss: 0.1090 - accuracy: 0.9666 - val_loss: 0.0892 - val_accuracy: 0.9744
Epoch 113/200
48000/48000 [==============================] - 3s 65us/sample - loss: 0.1112 - accuracy: 0.9662 - val_loss: 0.0895 - val_accuracy: 0.9746
Epoch 114/200
48000/48000 [==============================] - 3s 65us/sample - loss: 0.1082 - accuracy: 0.9668 - val_loss: 0.0894 - val_accuracy: 0.9742
Epoch 115/200
48000/48000 [==============================] - 3s 65us/sample - loss: 0.1068 - accuracy:

Epoch 163/200
48000/48000 [==============================] - 3s 62us/sample - loss: 0.0842 - accuracy: 0.9743 - val_loss: 0.0828 - val_accuracy: 0.9762
Epoch 164/200
48000/48000 [==============================] - 3s 62us/sample - loss: 0.0854 - accuracy: 0.9729 - val_loss: 0.0828 - val_accuracy: 0.9765
Epoch 165/200
48000/48000 [==============================] - 3s 62us/sample - loss: 0.0857 - accuracy: 0.9736 - val_loss: 0.0826 - val_accuracy: 0.9767
Epoch 166/200
48000/48000 [==============================] - 3s 63us/sample - loss: 0.0856 - accuracy: 0.9733 - val_loss: 0.0822 - val_accuracy: 0.9774
Epoch 167/200
48000/48000 [==============================] - 3s 64us/sample - loss: 0.0841 - accuracy: 0.9742 - val_loss: 0.0829 - val_accuracy: 0.9762
Epoch 168/200
48000/48000 [==============================] - 3s 66us/sample - loss: 0.0843 - accuracy: 0.9737 - val_loss: 0.0818 - val_accuracy: 0.9772
Epoch 169/200
48000/48000 [==============================] - 3s 63us/sample - loss: 0.08

In [14]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test Accuracy', test_acc)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

Test Accuracy 0.9774


### Dropout 의 추가로 인해 정확도 상승
- Test Accuracy : from 0.9645 to 0.9774